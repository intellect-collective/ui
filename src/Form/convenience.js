import { asField } from '../utils';

const Hidden = asField('value', { type: 'hidden' })('input');
const NumberField = asField('value', { type: 'number' })('input');
const Text = asField('value', { type: 'text' })('input');
const Textarea = asField('value')('textarea');
const Password = asField('value', { type: 'password' })('input');

export {
    Hidden,
    NumberField,
    Password,
    Text,
    Textarea
};
