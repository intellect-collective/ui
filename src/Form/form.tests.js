import React from 'react';
import { Form } from '../..';

describe('Form', () => {
    it('Missing onSubmit does not produce error', () => {
        const wrapper = mount(
            <Form action="/">
                <input type="submit" />
            </Form>
        );
        wrapper.find('input').simulate('submit');
    });

    it('Form submit calls handler', () => {
        const fn = jest.fn();
        const wrapper = mount(
            <Form action="/" handler={{ onSubmit: fn }}>
                <input type="submit" name="test" />
            </Form>
        );
        wrapper.find('input').simulate('submit');
        expect(fn).toHaveBeenCalledTimes(1);
        expect(fn.mock.calls[0][0]).toMatchObject({
            type: 'submit',
            target: {
                name: 'test'
            }
        });
    });

    it('Can set form encoding type', () => {
        const text = mount(<Form action="/" encType="text" />);
        expect(text.find('form').prop('encType')).toEqual('text/plain');
        const multipart = mount(<Form action="/" encType="multipart" />);
        expect(multipart.find('form').prop('encType')).toEqual('multipart/form-data');
    });

    it('Can set method to post', () => {
        const wrapper = mount(
            <Form action="/" method="post" />
        );
        expect(wrapper.find('form').prop('method')).toEqual('post');
    });

    it('Can set method to put', () => {
        const wrapper = mount(
            <Form action="/" method="put" overrideProperty="__method__" />
        );
        expect(wrapper.find('form').prop('method')).toEqual('post');
        expect(wrapper.find('input').prop('name')).toEqual('__method__');
        expect(wrapper.find('input').prop('value')).toEqual('put');
    });

    it('Can render component', () => {
        const fn = jest.fn(() => (<div />));
        mount(
            <Form action="/" component={ fn } />
        );
        expect(fn).toHaveBeenCalledTimes(1);
        expect(Object.keys(fn.mock.calls[0][0])).toIncludeSameMembers([
            'children',
            'getOriginal',
            'getLastValue',
            'getValue',
            'setValue',
            'getError',
            'setError',
            'isTouched',
            'setTouched',
            'isDirty',
            'getConflict'
        ]);
    });

    it('Can render function', () => {
        const fn = jest.fn(() => (<div />));
        mount(
            <Form action="/">{ fn }</Form>
        );
        expect(fn).toHaveBeenCalledTimes(1);
        expect(Object.keys(fn.mock.calls[0][0])).toIncludeSameMembers([
            'getOriginal',
            'getLastValue',
            'getValue',
            'setValue',
            'getError',
            'setError',
            'isTouched',
            'setTouched',
            'isDirty',
            'getConflict'
        ]);
    });
});
