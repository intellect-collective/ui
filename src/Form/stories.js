import React from 'react';
import PropTypes from 'prop-types';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withState } from '@dump247/storybook-state';
import {
    Checkbox, Field, Form, Password, Text
} from '../..';

class FormExample extends React.Component {
    static propTypes = {
        data: PropTypes.object,
        onChange: PropTypes.func
    };

    render() {
        return (
            <Form action="/" method="post" data={ this.props.data } onChange={ this.props.onChange } onSubmit={ (data, ev) => { ev.preventDefault(); ev.stopPropagation(); action('onSubmit')(data); } } className="form-horizontal">
                <p>Sign in for a better experience.</p>
                <Field name="email"
                        label="Email"
                        placeholder="Email"
                        className="form-group"
                        classNames={{
                            label: 'col-lg-2 control-label',
                            wrapper: 'col-lg-10',
                            field: 'form-control'
                        }}
                        wrapper="div"
                        component={ Text } />
                <Field name="password"
                        label="Password"
                        placeholder="Password"
                        className="form-group"
                        classNames={{
                            label: 'col-lg-2 control-label',
                            wrapper: 'col-lg-10',
                            field: 'form-control'
                        }}
                        wrapper="div"
                        component={ Password } />
                <div className="form-group">
                    <div className="col-lg-offset-2 col-lg-10">
                        <div className="i-checks">
                            <label>
                                <div className="icheckbox_square-green" style={{ position: 'relative' }}>
                                    <input type="checkbox" style={{ position: 'absolute', opacity: '0' }} />
                                    <ins className="iCheck-helper"
                                            style={{
                                                position: 'absolute',
                                                top: '0%',
                                                left: '0%',
                                                display: 'block',
                                                width: '100%',
                                                height: '100%',
                                                margin: '0px',
                                                padding: '0px',
                                                background: 'rgb(255, 255, 255)',
                                                border: '0px',
                                                opacity: '0'
                                            }} />
                                </div>
                                <i />
                                Remember me
                                <Checkbox name="remember" value="yes" />
                            </label>
                        </div>
                    </div>
                </div>
                <div className="form-group">
                    <div className="col-lg-offset-2 col-lg-10">
                        <button type="submit" name="go" className="btn btn-sm btn-white">Sign In</button>
                    </div>
                </div>
            </Form>
        );
    }
}

storiesOf('2 - Components / Form', module)
    .add('default', withState()(({ store }) => (<FormExample action="/" data={ store.state } onChange={ (ev) => store.set({ [ev.target.name]: ev.target.value}) } />)));
