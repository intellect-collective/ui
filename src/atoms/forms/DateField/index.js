import React from 'react';
import { format, isValid, toDate, parseISO } from 'date-fns';
import { asField } from '../utils';

const getValue = (raw) => {
	if (!raw) {
		return;
	}
	const iso = parseISO(raw);
	if (isValid(iso)) {
		return format(iso, 'yyyy-MM-dd');
	}
	return;
}

export const DateField = (props) => {
    return (<input { ...props } value={ getValue(props.value) } type="date" />);
};
DateField.displayName = 'DateField';

export { DateField as BasicDateField };

export default asField('value')(DateField);
