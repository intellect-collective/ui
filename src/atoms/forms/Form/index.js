import React from 'react';
import PropTypes from 'prop-types';

export const FormContext = React.createContext(null);

export default class Form extends React.Component {
    static propTypes = {
        action: PropTypes.string.isRequired,
        method: PropTypes.oneOf(['get', 'post', 'put']),
        overrideProperty: PropTypes.string,
        encType: PropTypes.oneOf(['text', 'multipart']),
        onChange: PropTypes.func,
        onSubmit: PropTypes.func,
        children: PropTypes.any,
        data: PropTypes.object
    };

    static defaultProps = {
        method: 'post',
        overrideProperty: '_method',
        data: {},
        onChange: () => {},
        onSubmit: () => {}
    };

    constructor(props) {
        super(props);
        this.getValue = this.getValue.bind(this);
        this.setValue = this.setValue.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.getEncType = this.getEncType.bind(this);
        this.getFormMethod = this.getFormMethod.bind(this);
        this.getMethodOverrideElement = this.getMethodOverrideElement.bind(this);
    }


    getValue(name) {
        if (this.props.data) {
            return this.props.data[name];
        }
        return undefined;
    }

    setValue(ev) {
        this.props.onChange(ev);
    }

    /**
     * Calls the onSubmit prop function, if provided, when the form is
     * submitted.
     * @param {Object} e - The React event
     */
    onSubmit(e) {
        if (this.props.onSubmit) {
            this.props.onSubmit(this.props.data, e);
        }
    }

    /**
     * Determines the encoding type of the form.
     * See https://www.ietf.org/rfc/rfc1867.txt
     * @returns {String} - The mime type to specify in th enctype attribute
     */
    getEncType() {
        if (this.props.encType === 'text') {
            return 'text/plain';
        }
        if (this.props.encType === 'multipart') {
            return 'multipart/form-data';
        }
        return this.props.encType;
    }

    /**
     * Determines the HTTP method to use when submitting the form.
     * See https://www.ietf.org/rfc/rfc1867.txt
     * @returns {String} - The http method to use
     */
    getFormMethod() {
        if (['get', 'post'].indexOf(this.props.method) !== -1) {
            return this.props.method;
        }
        return 'post';
    }

    /**
     * Produces a hidden element which provides the opportunity to emulate
     * http methods other than GET and POST. Requires that the server receiving
     * the request support method overriding.
     * @returns {Object} - A React element
     */
    getMethodOverrideElement() {
        if (['get', 'post'].indexOf(this.props.method) !== -1) {
            return null;
        }
        return (
            <input type="hidden"
                    name={ this.props.overrideProperty || '_method' }
                    value={ this.props.method } />
        );
    }

    /**
     * Builds the props that are to be passed through to the form element. We
     * need to get rid of some props that aren't supported by the form element,
     * so we take that opportunity here.
     * @returns {Object} - The final props object
     */
    getProps() {
        const props = {
            ...this.props,
            encType: this.getEncType(),
            method: this.getFormMethod(),
            onSubmit: this.onSubmit
        };
        delete props.overrideProperty;
        return props;
    }

    /**
     * Perform render
     * @returns {Object} - A React element
     */
    render() {
        const {
            children,
            onChange,
            onSubmit,
            ...props
        } = this.getProps();
        const ctx = {
            getValue: this.getValue,
            setValue: this.setValue
        };
        return (
            <form { ...props }>
                <FormContext.Provider value={ ctx }>
                    { this.getMethodOverrideElement() }
                    { this.props.children }
                </FormContext.Provider>
            </form>
        );
    }
}
