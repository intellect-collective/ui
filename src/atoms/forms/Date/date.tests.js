import React from 'react';
import DateField from '.';
import Form from '../Form';
import {
    basicChangeableTests,
    rerenderSuppressionTests
} from '../../../../test/utils';

const validProps = {
    name: 'test',
    value: '20120409',
    onClick: () => {}
};
const expectedProps = {
    name: 'test',
    value: '2012-04-09',
    onClick: expect.any(Function)
};

describe('Date', () => {
    basicChangeableTests(DateField, 'input', validProps, expectedProps);
    rerenderSuppressionTests(DateField, validProps);

    describe('Form Context Usage', () => {
        it('Gets value from form', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: '20140918' }}>
                    <DateField name="test" />
                </Form>
            );
            expect(wrapper.find('input').prop('value')).toEqual('2014-09-18');
        });
        it('Prefers local value', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: '20140918' }}>
                    <DateField name="test" value="20160514" />
                </Form>
            );
            expect(wrapper.find('input').prop('value')).toEqual('2016-05-14');
        });
    });
});
