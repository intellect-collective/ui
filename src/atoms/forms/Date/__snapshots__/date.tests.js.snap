// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date Should accept valid props 1`] = `
<FormContext
  name="test"
  onClick={[Function]}
  value="20120409"
>
  <Changeable
    name="test"
    onClick={[Function]}
    value="20120409"
  >
    <DateField
      name="test"
      onChange={[Function]}
      onClick={[Function]}
      value="20120409"
    >
      <input
        name="test"
        onChange={[Function]}
        onClick={[Function]}
        type="date"
        value="2012-04-09"
      />
    </DateField>
  </Changeable>
</FormContext>
`;
